bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/AbstractBigtableTable.java

```diff
@@ -276,15 +277,40 @@ public Result get(Get get) throws IOException {
   @Override
   public ResultScanner getScanner(final Scan scan) throws IOException {
     LOG.trace("getScanner(Scan)");
-    Span span = TRACER.spanBuilder("BigtableTable.scan").startSpan();
+    final Span span = TRACER.spanBuilder("BigtableTable.scan").startSpan();
     try (Scope scope = TRACER.withSpan(span)) {
 
       final ResultScanner scanner = clientWrapper.readRows(hbaseAdapter.adapt(scan));
       if (hasWhileMatchFilter(scan.getFilter())) {
         return Adapters.BIGTABLE_WHILE_MATCH_RESULT_RESULT_SCAN_ADAPTER.adapt(scanner, span);
       }
-      // TODO: need to end the span when stream ends
-      return scanner;
+
+      return new AbstractClientScanner() {
+        @Override
+        public Result next() throws IOException {
+          Result row = scanner.next();
+          if (row == null) {
+            // Null signals EOF.
+            span.end();
+            return null;
+          }
+          return row;
+        }
+
+        @Override
+        public void close() {
+          try {
+            scanner.close();
+          } finally {
+            span.end();
+          }
+        }
```

Comment:
This is a bit problematic: under normal usage this will end up calling span.end() twice. Also, when an exception is thrown, it needs to end the span with the appropriate status.

In general I actually think that trying to wrap a span around the lifetime of an iterator is problematic to begin with. I think we should probably remove this and ensure that the underlying ReadRowsOperation has a proper span

Reply:
refixed, see #2780